### PR TITLE
Read byte rewrite

### PR DESCRIPTION
**Pull request by R4PaSs**
As part of the stream refactor series, this patches cleanup msgpack a bit and introduce a lower-level `read_byte` implementation on which `read_byte(s)` will rely for reading data from a source.

In the future, the lowest-level interface to read bytes, and the only one required for implementing a new stream will be the `raw_read_byte(s)` functions.

A cleanup of the current streams will be needed for this to be used properly, coming in the next few PRs.

Note: The first two commits have already been submitted under #2632, therefore disregard them for this PR.

### CODE REVIEW

File: lib/core/file.nit [hunk 1]
```diff
@@ -511,8 +511,8 @@ class Path
 		var output = dest.open_wo
 
 		while not input.eof do
-			var buffer = input.read_bytes(1024)
-			output.write_bytes buffer
+			var buf = input.read_bytes(4096)
+			output.write_bytes buf
```
Comment (privat) [inline lib/core/file.nit]: why did you rename `buffer` to `buf`?

File: lib/core/stream.nit [hunk 2]
```diff
@@ -95,32 +95,71 @@ end
...
-	fun read_byte: Int is abstract
+	fun read_byte: Int do
+		var llen = lookahead_length
+		if llen == 0 then return raw_read_byte
+		var lk = lookahead
+		var b = lk[0].to_i
+		if llen == 1 then
+			lookahead_length = 0
+		else
+			lk.lshift(1, llen - 1, 1)
```
Comment (privat) [inline lib/core/stream.nit]: ouch! Shouldn't a cursor better that shifting the whole lookahead buffer?

File: lib/core/stream.nit [hunk 3]
```diff
@@ -95,32 +95,71 @@ end
...
+
+	# Reads up to `max` bytes from source and stores them in `bytes`
+	fun read_bytes_to_cstring(bytes: CString, max: Int): Int do
+		var llen = lookahead_length
+		if llen == 0 then return raw_read_bytes(bytes, max)
+		var rd = max.min(llen)
+		var lk = lookahead
+		lk.copy_to(bytes, rd, 0, 0)
+		if rd < llen then
+			lk.lshift(rd, llen - rd, rd)
```
Comment (privat) [inline lib/core/stream.nit]: same

File: lib/core/stream.nit (hunk 2 above)
Comment (R4PaSs) [reply lib/core/stream.nit]: Yeah, that would be better indeed. But, since the codec will need a sequence of bytes to determine if the chars to read are valid, I don't have much of a choice.
Besides, this should be a very rare case as it will only happen when an invalid char was read (which, if you chose your encoding properly should not happen that much), and the lookahead should not be too long (4 bytes max in the case of unicode).

File: lib/core/file.nit (hunk 1 above)
Comment (R4PaSs) [reply lib/core/file.nit]: Artefacts of reworking it all, I'll leave the `buffer` name

**Comment by R4PaSs** [comment]
Agreed on the msgpack commit, it deserves to be split in another PR; I'll open it soon.

File: lib/core/stream.nit [hunk 4]
```diff
@@ -798,11 +836,10 @@ class BytesReader
 
 	redef fun close do bytes = new Bytes.empty
 
-	redef fun read_all_bytes
```
Comment (xymus) [inline lib/core/stream.nit]: Why remove `read_all_bytes`?

File: lib/msgpack/read.nit [hunk 5]
```diff
@@ -70,83 +70,58 @@ redef class Reader
...
+
+		else if typ == 0xDC or typ == 0xDD then
+			# array16 and array32
+			var len_ln = 2 << (typ - 0xDC)
+			var lenbuf = read_bytes(len_ln)
+			return read_msgpack_array_data(lenbuf.to_i)
+
+		else if typ == 0xDE or typ == 0xDF then
+			# map16 and map32
+			var len_ln = 2 << (typ - 0xDC)
```
Comment (xymus) [inline lib/msgpack/read.nit]: 0xDE?

File: lib/core/stream.nit [hunk 6]
```diff
@@ -65,35 +95,71 @@ end
 abstract class Reader
 	super Stream
 
-	# Decoder used to transform input bytes to UTF-8
-	var decoder: Codec = utf8_codec is writable
+	# Read a byte directly from the underlying stream, without
+	# considering any eventual buffer
```
Comment (xymus) [inline lib/core/stream.nit]: Be careful with line wraps on titles, AFAIK the doc tools use only the first line.

File: lib/core/stream.nit [hunk 7]
```diff
@@ -65,35 +95,71 @@ end
...
-		var s = new CString(i)
-		var buf = new Bytes(s, 0, i)
-		while i > 0 and not eof do
-			var c = read_byte
-			if c < 0 then
-				continue
-			end
-			buf.add c.to_b
-			i -= 1
+	# Reads up to `max` bytes from source
```
Comment (xymus) [inline lib/core/stream.nit]: When does it not read `max` chars? Only when at EOF or also when data is not yet available?

Comment (R4PaSs) [reply lib/core/stream.nit]: That will depend on the implementation of `read_bytes_to_cstring`, I'd think it would be better to have a C-like semantic, i.e. reading up-to-`max` bytes if available, or the available buffered amount, while blocking on the first read if none are available.

File: lib/core/stream.nit (hunk 4 above)
Comment (R4PaSs) [reply lib/core/stream.nit]: The super implementation should do the job alright in the first place, no need to redefine it here I'd think.
But come to think of it, in this particular case, it can produce a significant speedup, I'll leave the previous implementation available for this.

File: lib/msgpack/read.nit (hunk 5 above)
Comment (R4PaSs) [reply lib/msgpack/read.nit]: Yep, that would explain why the tests fail here, thanks!

**Comment by R4PaSs** [comment]
Split msgpack commit into #2640, fixed the removed `read_all_bytes` in `BytesReader`
